File: 3.1.JavaConsumerProducerAPI/src/main/java/org/ogomez/nontx/SimpleProducer.java
package org.ogomez.nontx;

import java.util.Random;
import java.util.concurrent.ExecutionException;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class SimpleProducer {

  public static void main(String[] args) throws ExecutionException, InterruptedException {

    //instancia de Logger solo para poder pintar y facilitar el seguimiento de nuestro proceso
    final Logger logger = LoggerFactory.getLogger(SimpleProducer.class);

    //creamos una instnacia de nuestro producer usando la clase de configuración que hemos
    //creado para esto.
    KafkaProducer<String, String> producer = KafkaConfig.createKafkaProducer();

    //Clase de utilidad java para generar aleatorios
    Random random = new Random();

    //bucle de mil iteraciones que usaremos para generar mensajes aleatorios
    for (int i = 0; i < 1000; i++) {

      //Generamos una key aleatorio usando Random, para poder simular el balanceo por una partition key
      String key = String.valueOf(random.ints(0, 9).findFirst().orElse(0));
      //mensaje en un string en formato json (en realidad valdria cualquier string ya que no estamos serializando con ningun esquema)
      //con el numero de iteracion que genero el mensaje (atentos a la asincronia al producir)
      String value = "{\"msg\":\"" + i + "\"}";

      //Producer record, parametrizado con los mismos tipos que nuestro KafkaProducer. Le pasamos
      // 3 valores: Topic a producir, key por la que particionar (recordemos en caso de no pasarla
      //particionaria por round robin), y el value o mensaje a producir
      ProducerRecord<String, String> record =
          new ProducerRecord<>(KafkaConfig.INPUT_TOPIC, key, value);

      logger.info("Key: " + key);
      logger.info("value:" + value);

      //Enviamos el mensaje al broker:
      //    - Notar que el metodo devuelve los metadatos del record producido, no el metodo en si
      //    - Este metodo devuelve un Futuro, esto quiere decir que es un proceso asincrono, send
      //      enviara el mensaje al broker y quedara esperando la respuesta de este sin bloquear el
      //      hilo de ejecución.
      //    - Por esta naturaleza de "Futuro" la funcion send recibe como parametro una  funcion
      //      de callback es decir le decimos como tratar la respuesta, tratando un RecordMetadata
      //      en caso de ir bien, y una excepcion si falla.
      producer.send(record, (recordMetadata, e) -> {
        if (e == null) {
          logger.info("Received new metadata. \n" +
              "Topic:" + recordMetadata.topic() + "\n" +
              "Partition: " + recordMetadata.partition() + "\n" +
              "Offset: " + recordMetadata.offset() + "\n" +
              "Timestamp: " + recordMetadata.timestamp());
        } else {
          logger.error("Error while producing", e);
        }
      }).get();
    }

    //limpiamos la cola del productor (forzamos a que los envie a Kafka, resolviendo los futuros)
    // antes de cerrar la conexion
    producer.flush();
    //cerramos la conexion con el broker.
    producer.close();

  }
}

File: 3.2.PythonConsumerProducerAPI/simpleProducer.py
#!/usr/bin/env python

import json
import logging
import random

# importamos de la libreria kafka KafkaProducer
from kafka import KafkaProducer

log = logging.getLogger(__name__)

# Creamos nuestro Kafka Producer pasandole TODOS los Broker
# Utilizaremos la libreria JSON para serializar nuestros mensajes
producer = KafkaProducer(
    bootstrap_servers=['127.0.0.1:9092', '127.0.0.1:9093', '127.0.0.1:9094'],
    value_serializer=lambda m: json.dumps(m).encode('utf-8'))

# Funcion para el caso de exito en la produccion del metodo
def on_send_success(record_metadata):
  print(record_metadata.topic)
  print(record_metadata.partition)
  print(record_metadata.offset)

# que haremos en caso de error
def on_send_error(ex):
  log.error('I am an Error', exc_info=ex)
  # handle exception


# prducimos 1000 mensajes en este caso no particionamos
for i in range(1000):
  key = str(random.randint(0, 9))
  producer.send('simple-topic', key=key.encode('utf-8'),
                value={'msg': str(i)}).add_callback(
      on_send_success).add_errback(on_send_error)

producer.flush()
